GoogleDataTransport/GDTCORLibrary/GDTCORUploadPackage.m

```diff
@@ -31,7 +31,7 @@ @implementation GDTCORUploadPackage {
   BOOL _isHandled;
 
   /** A timer that will regularly check to see whether this package has expired or not. */
-  NSTimer *_expirationTimer;
+  __weak NSTimer *_expirationTimer;
```

Comment:
This can't be the fix for this issue. This reference is owned by `GDTCORUploadPackage` and is the only reference to the timer. Making this weak will make it subject to immediate collection by ARC and will effectively disable the timer. This doesn't break `GDTCORUploadPackage` because nil dispatch will continue to succeed.

Reply:
The timer has a reference with `GDTCORUploadPackage` from the initialization of the `NSTimer` I only make the timer with nil after setting it as invalidate  and I am sure the `GDTCORUploadPackage` will be copied with `[self copy]`

Reply:
Wouldn't `expirationTimer` be immediately `nil` after assigning it to the weak variable on line 44, regardless of copies?

Reply:
<img width="745" alt="Screen Shot 2020-04-14 at 9 34 52 PM" src="https://user-images.githubusercontent.com/1476754/79261452-85044600-7e98-11ea-8fbd-658558da3a0f.png">

 The initialization of NSTimer will make retain cycle with this style

Reply:
I grant that there's a retain cycle, but the expirationTimer needs to be held strongly by `GDTCORUploadPackage` or it will be released essentially immediately. There needs to be a strong reference to the timer.

The problem is that `NSTimer` needs to keep a weak reference to the `GDTCORUploadPackage` instance.

This would be better written without using selectors.

```objc
__weak typeof(self) weakSelf = self;
_expirationTimer = [NSTimer scheduledTimerWithTimeInterval:5.0
                                                   repeats:YES
                                                   block:^(NSTimer *timer) {
  __strong typeof(self) strongSelf = weakSelf;
  if (strongSelf) {
    [strongSelf checkIfPackageIsExpired:timer];
  }
});
```

Reply:
Wil's suggestion is the correct way of doing this.

Reply:
it's not the solution by the way [scheduledtimerwithtimeinterval](https://developer.apple.com/documentation/foundation/nstimer/2091889-scheduledtimerwithtimeinterval?language=objc
) it's support **iOS 10+** 
and  `GoogleDataTransport` support from **iOS 8** depend [GoogleDataTransport.podspec](https://github.com/firebase/firebase-ios-sdk/blob/master/GoogleDataTransport.podspec)

Reply:
I try this solution 
```
    __weak id weakSelf = self;
    _expirationTimer = [NSTimer scheduledTimerWithTimeInterval:5.0
                                                        target:weakSelf
                                                      selector:@selector(checkIfPackageIsExpired:)
                                                      userInfo:nil
                                                       repeats:YES];
```
but unfortunately doesn't work also

<img width="1677" alt="Screen Shot 2020-04-15 at 11 29 39 AM" src="https://user-images.githubusercontent.com/1476754/79316212-b4ed3100-7f0c-11ea-8e47-25caeddbdb53.png">

Reply:
How I can check  the expire packet scenario ?

Reply:
Passing a weak reference into the timer's `target:` argument won't cause the target to not be strongly retained, since ARC can't conditionally remove `retain`s from the body of the timer method (but it can avoid inserting `retain`s when constructing a block literal with a weak reference).

The memory leak is caused by the timer not being invalidated in some cases. Gil's fix resolves this by weakening the timer's reference to `self`, so when `self` is deallocated the timer will be invalidated in the `dealloc` method. The best way to achieve the same thing in iOS versions below 10 is to use a helper object that holds a weak reference to the GDTCORUploadPackage instance, and then passing that object to the timer instead of `self`.

I'm not sure how to check the packet expiration method, but maybe unit tests would be sufficient for testing this leak.

Reply:
@mikehaney24 Which solution will you implement the one that uses block?